src/hotspot/cpu/x86/stubGenerator_x86_64_fmod.cpp

```diff
@@ -0,0 +1,545 @@
+
+//   if (a < b)  return x + sgn_a;
+  __ ucomisd(xmm3, xmm4);
+  __ jcc(Assembler::belowEqual, L_104a);
+  __ vaddsd(xmm0, xmm2, xmm0);
+  __ jmp(L_11bd);
+
+//   if (((mxcsr & 0x6000)!=0x2000) && (a < b * 0x1p+260))
+  __ bind(L_104a);
+  __ andl(rax, 0x6000);
+  __ cmpl(rax, 0x2000);
+  __ jcc(Assembler::equal, L_10c1);
+  __ vmulsd(xmm0, xmm3, ExternalAddress((address)CONST_1p260), rax);
+  __ ucomisd(xmm0, xmm4);
+  __ jcc(Assembler::belowEqual, L_10c1);
+//   {
+//     q = DP_DIV(a, b);
+  __ vdivpd(xmm0, xmm4, xmm3, Assembler::AVX_128bit);
+//     corr = DP_SHR(DP_FNMA(b, q, a), 63);
+  __ movapd(xmm1, xmm0);
+  __ vfnmadd213sd(xmm1, xmm3, xmm4);
+  __ movq(xmm5, xmm1);
+  __ vpxor(xmm1, xmm1, xmm1, Assembler::AVX_128bit);
+  __ vpcmpgtq(xmm5, xmm1, xmm5, Assembler::AVX_128bit);
+//     q = DP_PSUBQ(q, corr);
+  __ vpaddq(xmm0, xmm5, xmm0, Assembler::AVX_128bit);
+//     q = DP_TRUNC(q);
+  __ vroundsd(xmm0, xmm0, xmm0, 3);
+//     a = DP_FNMA(b, q, a);
+  __ vfnmadd213sd(xmm0, xmm3, xmm4);
+  __ align32();
+//     while (b <= a)
+  __ bind(L_1090);
+  __ ucomisd(xmm0, xmm3);
+  __ jcc(Assembler::below, L_11b9);
+//     {
+//       q = DP_DIV(a, b);
+  __ vdivsd(xmm4, xmm0, xmm3);
+//       corr = DP_SHR(DP_FNMA(b, q, a), 63);
+  __ movapd(xmm5, xmm4);
+  __ vfnmadd213sd(xmm5, xmm3, xmm0);
+  __ movq(xmm5, xmm5);
+  __ vpcmpgtq(xmm5, xmm1, xmm5, Assembler::AVX_128bit);
+//       q = DP_PSUBQ(q, corr);
+  __ vpaddq(xmm4, xmm5, xmm4, Assembler::AVX_128bit);
+//       q = DP_TRUNC(q);
+  __ vroundsd(xmm4, xmm4, xmm4, 3);
+//       a = DP_FNMA(b, q, a);
+  __ vfnmadd231sd(xmm0, xmm3, xmm4);
+  __ jmp(L_1090);
+//     }
+//     return DP_XOR(a, sgn_a);
+//   }
+
+//   __asm { ldmxcsr DWORD PTR [mxcsr_rz] }
+  __ bind(L_10c1);
+  __ ldmxcsr(Address(rsp, 0x04));
+
+//   q = DP_DIV(a, b);
+  __ vdivpd(xmm0, xmm4, xmm3, Assembler::AVX_128bit);
+//   q = DP_TRUNC(q);
+  __ vroundsd(xmm0, xmm0, xmm0, 3);
+
+//   eq = TRANSFER_HIGH_INT32(q);
+  __ extractps(rax, xmm0, 1);
+
+//   if (__builtin_expect((eq >= 0x7fefffffu), (0==1))) goto SPECIAL_FMOD;
+  __ cmpl(rax, 0x7feffffe);
+  __ jcc(Assembler::above, L_10e7);
+
+//   a = DP_FNMA(b, q, a);
+  __ vfnmadd213sd(xmm0, xmm3, xmm4);
+  __ jmp(L_11af);
+
+// SPECIAL_FMOD:
+
+//   // y==0 or x==Inf?
+//   if ((b == 0.0) || (!(a <= DP_CONST(7fefffffffffffff))))
+  __ bind(L_10e7);
+  __ vpxor(xmm5, xmm5, xmm5, Assembler::AVX_128bit);
+  __ ucomisd(xmm3, xmm5);
+  __ jcc(Assembler::notEqual, L_10f3);
+  __ jcc(Assembler::noParity, L_111c);
+
+  __ bind(L_10f3);
+  __ movsd(xmm5, ExternalAddress((address)CONST_MAX), rax);
+  __ ucomisd(xmm5, xmm4);
+  __ jcc(Assembler::below, L_111c);
+//     return res;
+//   }
+//   // y is NaN?
+//   if (!(b <= DP_CONST(7ff0000000000000))) {
+  __ movsd(xmm0, ExternalAddress((address)CONST_INF), rax);
+  __ ucomisd(xmm0, xmm3);
+  __ jcc(Assembler::aboveEqual, L_112a);
+//     res = y + y;
+  __ vaddsd(xmm0, xmm0, xmm1);
```

Comment:
Should this be: __ vaddsd(xmm0, xmm1, xmm1);

Reply:
This is correct as written according to the disassembly.

Reply:
After further investigation, you are correct.  Changed.